packages/loopring_v3/contracts/impl/agents/FastWithdrawalAgent.sol

```diff
@@ -73,22 +59,18 @@ contract FastWithdrawalAgent is ReentrancyGuard
         address from;                   // The owner of the account
         address to;                     // The address that will receive the tokens withdrawn
         address token;
-        uint    amount;
+        uint96  amount;
```

Comment:
@Brechtpd I'd like to make sure I understand this contract. Let me an example,

Alice wants to with some token, she signs an offchain withdrawal as follows
```
withdrawal#1:  {owner: aliceAddr, to: aliceAddr2, amount: 10, token: LRC, sig: eddsa_sig}
```

The relayer checks Alice has >= 10 LRC and freeze 10 LRC.

The relayer finds liquidator provider Bob has more than 10 LRC, the operator then notifies Bob to executeFastWithdrawal. In this function, Bob transfers 10 LRC from `bobAddr` to `aliceAddr2`, and created a withdrawal approval hash onchain to authorize a NEW withdrawal from `aliceAddr` to `bobAddr` without EdDSA signature, then send this 

```
  withdrawal#2: {owner: aliceAddr, to: bobAddr, amount: 10, token: LRC, sig: ""}
```

Note that:
1. withdrawal#1 is still valid but will be dropped by the relayer (if the relayer is evil, he can still use this request)
2. withdrawal#2 can be created by the operator or the liquidity provider , but if this request is forever rejected, the token will not be withdrawable by the liquidity provider Bob, but Alice, in withdrawal mode.

I think both 1 and 2 imply some concerns.

Please let me know if my understanding is correct.

Reply:
Already discussed, but will briefly answer here as well for reference.

> withdrawal#1 is still valid but will be dropped by the relayer (if the relayer is evil, he can still use this request to double-spend Alice's money to charlyAddr.

Alice only signs the onchain fast withdrawal, no offchain withdrawal so no double-spending.

> This means the liquidity provider does have to trust the operator.

Yes.

> A problem that may occur is a race condition. What if Bob is trying to executeFastWithdrawal while the operator decided to process Alice's original withdrawal request? It seems there is no way to prevent both of them from being done.

Liquidity provider is never sure he will get his money back, the operator controls what transactions happen and can either never the funds back or let Alice spend the funds differently.


This method is still the most flexible way to do things, but for a simple fast withdrawal it seems we can make it so that offchain withdrawal requests can be used directly as well.

Reply:
The method to allow fast withdrawals on all withdrawals is less efficient because it needs 2 onchain token transfers. The 1st one for the fast withdrawal, the 2nd to do the withdrawal to the liquidity provider.

The separate fast withdrawal only needs 1 onchain transfer, because the liquidity provider gets paid back on layer 2 with a transfer.

Ideally we could also execute the withdrawal request as a transfer, but that would probably require some strange logic so probably not really something we want to do.

A fast withdrawal (with redirected withdrawal) will cost ~35% more gas than a fast withdrawal (with transfer), 60,000 gas vs 45,000 gas. And ~200% more gas than just a normal withdrawal, ~20,000 gas.

Reply:
- Redirect-based fast withdrawal takes two onchain transactions, one from the liquidity provider to the user, then the deposit contract to the liquidity provider.
- The current fast withdrawal also needs two onchain transactions, one from the user as a request, then then one from the liquidity provider to the user. 

It's a matter of who pays the fees.

Maybe we can do the following:

- the user submits an offchain withdrawal
- the liquidity provider send fund on layer 1 to the deposit's destination address, in the same tx, create an approval hash that correspond to a transfer from the user to the liquidity provider's account on layer2.

Reply:
> The current fast withdrawal also needs two onchain transactions, one from the user as a request, then then one from the liquidity provider to the user.

Not really, the user just provides a signature which only costs ~4,000 gas, and the signature check is done as part of the tx of the fast withdrawal so no overhead of an extra Ethereum tx. We do save around ~15,000 gas because of that. Both do need to store data inside the exchange contract so yeah same costs there. 

> Maybe we can do the following:

If we only have an EdDSA signature for the withdrawal we have to use that somehow to authorize any transaction that happens as part of the fast withdrawal, otherwise the agent contract can basically approve whatever transaction it wants. That's why we use the signed fast withdrawal request in the current Agent implementation, that's a cheap and easy way to do it. We also have to watch out not to use more storage than needed, storing a hash will cost 20,000 gas each!

So somehow we need to validate the EdDSA withdrawal request and make it into a transfer. Perhaps the withdrawal could be validated without state changes. When the withdrawal is processed onchain we see a fast withdrawal was done for it and now a transfer is approved instead and that transfer can then be used by the operator... but that seems pretty messy. Can't really think of anything better now. We could also make a circuit that processes the withdrawal and turns it into a transfer authorized by a fast withdrawal but that seems a bit overkill.